app/src/main/java/com/eveningoutpost/dexdrip/receivers/aidex/AidexReceiver.java

```diff
@@ -0,0 +1,297 @@
+        BgReading.bgReadingInsertFromInt(bgValueMgDl, timeStamp, segmentation_timeslice, false);
+
+    }
+
+
+    private void checkIfCorrectSensorIsRunning(String sensorId, long timeStamp) {
+        Sensor currentSensor = Sensor.currentSensor();
+        if(currentSensor!=null) {
+            if (!currentSensor.uuid.equals(sensorId)) {
+                if (JoH.pratelimit("aidex-sensor-restart", 1200)) {
+                    Sensor.stopSensor();
+                    Sensor.create(timeStamp - 10000, sensorId);
+                }
+            }
+        } else {
+            Sensor.create(timeStamp - 10000, sensorId);
+        }
+    }
+
+
+    private void processCalibration(Bundle bundle) {
+        if (Pref.getBooleanDefaultFalse("accept_broadcast_calibrations")) {
+
+            final long calibration_timestamp = bundle.getLong(AIDEX_TIMESTAMP, -1);
+            double bgValue = bundle.getDouble(AIDEX_BG_VALUE, -1);
+            final String units = bundle.getString(AIDEX_BG_TYPE, AidexBroadcastIntents.UNIT_MG_DL);
+
+            final long timeoffset = JoH.tsl() - calibration_timestamp;
+
+            if (bgValue > 0) {
+
+                if (timeoffset < 0) {
+                    Home.toaststaticnext(gs(R.string.got_calibration_in_the_future__cannot_process));
+                    return;
+                }
+
+                double valueMgDl = bgValue;
+
+                if (AidexBroadcastIntents.UNIT_MMOL_L.equals(units)) {
+                    valueMgDl = bgValue * Constants.MMOLL_TO_MGDL;
+                }
+
+                UserError.Log.ueh(TAG, "Adding blood test calibration: " + JoH.qs(bgValue, 2) + " offset ms: " + JoH.qs(timeoffset, 0));
+
+                BloodTest.create(calibration_timestamp, valueMgDl, "Aidex", null);
+            } else {
+                Log.e(TAG, "Received broadcast calibration without glucose number");
+            }
+        } else {
+            Log.w(TAG, "Received broadcast calibration from Aidex, but we don't accept calibrations (check settings)");
+        }
+    }
+
+
+    private void processSensorStart(Bundle bundle) {
+        if (bundle != null && bundle.containsKey(AIDEX_TIMESTAMP)) {
+            long sensorStartTime = bundle.getLong(AIDEX_TIMESTAMP);
+            String sensorId = bundle.getString(AIDEX_SENSOR_ID);
+
+            if (stopCurrentSensor(sensorId)) {
+                Sensor.create(sensorStartTime, bundle.getString(AIDEX_SENSOR_ID));
+            }
+        }
+    }
+
+    private boolean stopCurrentSensor(String nextSensorId) {
+        Sensor last = Sensor.currentSensor();
+        if (last != null) {
+            if (last.uuid.equals(nextSensorId)) {
+                // if we restart sensor we would have same uuid, so we don't do anything.
+                return false;
+            }
+            Sensor.stopSensor();
+        }
+
+        return true;
+    }
+
+
+    private void processSensorRestart(Bundle bundle) {
```

Comment:
@tolot27 this is the function I was referring to.

Reply:
Is rate limiting required in `checkIfCorrectSensorIsRunning`? Maybe, @andyrozman misunderstood the place where to implement it.

Reply:
Actually processSensorRestart would be called only if we received this event from their side, which *might* happen when they restart sensor (I am not sure if they implemented it that way)... But method checkIfCorrectSensorIsRunning() is called on each New_BG event, which is sent every 5 minutes. 
So checkIfCorrectSensorIsRunning() is the one that must be rate limited...